### ISSUE TITLE
JS

### ISSUE THREAD

**Issue by Holayn** [Holayn/knowledge#2]
- become intimately familiar with JS
- be able to answer interview questions regarding JS

**Comment by Holayn** [comment]
### Event queue

- events created are placed into queue
- engine periodically looks in queue to see if anything there ONLY if the execution stack is empty
  - creates execution context for item in queue, does for all item

### Event loop
- the mechanism that checks event queue (callback and promise queues) when main call stack is empty.
- [good article explaining](https://medium.com/@Rahulx1/understanding-event-loop-call-stack-event-job-queue-in-javascript-63dcd2c71ecd)
  - if call stack empty (i.e. your normal sync code), engine will look in promises queue and execute everything there, then goes to callback queue
  - sync code called, async code only is called after sync code in main stack is done
    - all code is pushed to this main stack and is executed as such. async stuff gets put into event queue.
- good vid: https://2014.jsconf.eu/speakers/philip-roberts-what-the-heck-is-the-event-loop-anyway.html

### Event loop blocking
- example: express server - a route that has a callback that has a long job will block all other routes from handling requests
  - https://medium.com/@witkesam/solving-the-blocked-event-loop-in-node-js-abb6cac281a7
  - deep dive: https://snyk.io/blog/nodejs-how-even-quick-async-functions-can-block-the-event-loop-starve-io/
- https://nodejs.org/es/docs/guides/dont-block-the-event-loop/

**Comment by Holayn** [comment]
### Scope and scope chain

- every execution context has a reference to its outer environment
  - outer environment = the lexical environment where a function is defined, where the function definition sits environment-lexically
  - if a variable isn't in the execution environment, it would go to the outer environment


#### Functions

- `this` refers to global execution context
- when a function is in an object, `this` refers to that object

```
var c = {  
  log: function() {
    console.log(this); // c
  },
  log2: function() {
    var d = function() { console.log(this) }
    d(); // context where c.log2() is called
  },
  log3: function() {
    var self = this;
    console.log(self);
    var d = function() { console.log(self) }
    d(); 
  }
}
```

#### If statements
They don't create new scope
```
if (true) {
  // this if statement doesn't create a new scope
  // so varInIf is available in the global scope
  var varInIf = 'inside an if statement';
}
console.log(varInIf);  // logs 'inside an if statement'
```

`let` is block scoped so won't run into this problem
`var` is only scoped to functions

**Comment by Holayn** [comment]
### Functions

- don't forget that they're just objects - first-class citizens!
- functions in objects are called ***methods***

#### When a function is called...

- a new execution context is created and put onto the execution stack. They have their own `this` context
- even when an execution context is removed, its variable environment / memory space remains (until garbage collections removes it)
- execution context contains information about where the function was called from (the call-stack), how the function was invoked, what parameters were passed, etc. One of the properties of this record is the `this` reference which will be used for the duration of that function's execution.

- `this` is a binding made for each function invocation, based entirely on its call-site (how the function is called) - what object is calling it!
- Finding the call-site is generally: "go locate where a function is called from", but it's not always that easy, as certain coding patterns can obscure the true call-site. What's important is to think about the call-stack (the stack of functions that have been called to get us to the current moment in execution). The call-site we care about is in the invocation before the currently executing function.

resources: [link](https://github.com/getify/You-Dont-Know-JS/blob/1st-ed/this%20%26%20object%20prototypes/ch2.md)
```
`function foo() {
	console.log( this.a );
}

var obj = {
	a: 2,
	foo: foo
};

obj.foo(); // 2
```

```
function foo() {
	console.log( this.a );
}

var obj = {
	a: 2,
	foo: foo
};

var bar = obj.foo; // function reference/alias!

var a = "oops, global"; // `a` also property on global object

bar(); // "oops, global"
```

```
function foo() {
	console.log( this.a );
}

function doFoo(fn) {
	// `fn` is just another reference to `foo`

	fn(); // <-- call-site!
}

var obj = {
	a: 2,
	foo: foo
};

var a = "oops, global"; // `a` also property on global object

doFoo( obj.foo ); // "oops, global"
```

#### IIFEs
- code in an IIFE won't collide with other code written, since vars in function are in its own execution context

#### `call()`, `apply()`, `bind()`
- changing the `this` context when the execution context is created for the function

**Comment by Holayn** [comment]
### Closures

Functions will always have access to the variables it's supposed to have access to

![image](https://user-images.githubusercontent.com/3732034/90990586-ab39a100-e570-11ea-8387-927277737198.png)

- new execution contexts are created whenever the same function is run, forming their own closures (i.e. factory functions)

You will have closures whenever having function declarations. "Closing" over variables so that have access to them when calling the function. 
- i.e. creating a function that returns another function

Useful for:
- data privacy (i.e. function returning another function/object with methods)
- callbacks

**Comment by Holayn** [comment]
### Prototype

JS has prototypal inheritance involving a **prototype chain**

- objects have properties and methods
- objects have prototype property called `__proto__` that points to the prototype of that object (i.e. the parent)
- if a property does not exist in an object, it goes to check in the `__proto__` - if not there, check its `__proto__`, etc
(read more about object links [here](https://github.com/getify/You-Dont-Know-JS/blob/1st-ed/this%20%26%20object%20prototypes/ch5.md#object-links))

Everything is an object (except primitives) and they all have a prototype (except the ultimate JS base `Object` which is the bottom of the proto-chain.
- This makes all objects have access to `toString()` for example

`instanceof` goes up the `__proto__` chain to see if in it

This is **different from the scope chain!**

#### prototype vs __proto__
Former is property of class constructor, latter is property of class instance

`__proto__` is an object in every class instance that points to the prototype it was created from.

> `iPhone.prototype` provides a blueprint for building an iPhone, `newPhone.__proto__` affirms that the iPhone has indeed been built according to that specific blueprint

[source](https://medium.com/javascript-in-plain-english/proto-vs-prototype-in-js-140b9b9c8cd5)

Constructors have a special property whose unfortunate name is `prototype`. The value of the property becomes the prototype of the instances created by that constructor.
```
function Foo() {}
console.log(Object.getPrototypeOf(new Foo()) === Foo.prototype); // true

var obj = new MyConstructorFunction();
obj.__proto__ === MyConstructorFunction.prototype // true
```

[source](https://www.reddit.com/r/javascript/comments/49xaze/proto_vs_prototype/)


Note: `__proto__` is a recent addition to the language (ES6). `[[Prototype]]` is what's referenced in the spec.

```
Object.defineProperty( Object.prototype, "__proto__", {
	get: function() {
		return Object.getPrototypeOf( this );
	},
	set: function(o) {
		// setPrototypeOf(..) as of ES6
		Object.setPrototypeOf( this, o );
		return o;
	}
} );
```

---
Side notes: 
- You don't create multiple instances of a class. **You can create multiple objects that [[Prototype]] link to a common object. But by default, no copying occurs**, and thus these objects don't end up totally separate and disconnected from each other, but rather, quite linked.
  - Links between objects
  - Using `Object.create()` to reliably create links

[good tldr](https://github.com/getify/You-Dont-Know-JS/blob/1st-ed/this%20%26%20object%20prototypes/ch5.md#review-tldr)

**Comment by Holayn** [comment]
#### `strict` mode
Tells the parser to be stricter, clean errors instead of hiding them
- must declare variables in order to use them
- can't delete things that can't be deleted
- can't write to readonly vars

Changes `this` behavior inside of functions - see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this)

a whole host of other things!

Makes it impossible to accidentally create global variables.
Makes assignments which would otherwise silently fail to throw an exception.
Makes attempts to delete undeletable properties throw (where before the attempt would simply have no effect).
Requires that function parameter names be unique.
this is undefined in the global context.
It catches some common coding bloopers, throwing exceptions.
It disables features that are confusing or poorly thought out.

**Comment by Holayn** [comment]
### Hoisting

The execution context has two phases: the creation phase and the execution phase.

Creation phase
- Recognize declared functions and vars
- Sets up memory space for variables and functions (**hoisting**)
  - functions and variables allocated space in memory
  - entire function is put into memory (including code)
  - undefined placeholders are used, as variable values are determined at runtime. All variables in JS are initially set to undefined!

Execution phase
- runs written code line by line

Example case of seeing this happening:
```
var text = 'outside';
function logIt(){
    console.log(text);
    var text = 'inside';
};
logIt(); // logs undefined since variable DECLARATIONS are hoisted

// really looks like this:
var text;
console.log(text);
text = 'inside'
```

Function declarations are hoisted over variable declarations but not over variable assignments

Readings: [link](https://www.digitalocean.com/community/tutorials/understanding-hoisting-in-javascript#:~:text=Hoisting%20is%20a%20JavaScript%20mechanism,scope%20is%20global%20or%20local.)

Variable assignment over function declaration
```
var double = 22;

function double(num) {
  return (num*2);
}

console.log(typeof double); // Output: number
```
Function declarations over variable declarations
```
var double;

function double(num) {
  return (num*2);
}

console.log(typeof double); // Output: function
```

BASICALLY
- use `let` to avoid these weird cases - you can't reference `let` declared vars before they're declared

**Comment by Holayn** [comment]
# Arrow functions

They don't have their own `this` pointing to their execution context

Uses lexical scope to resolve `this` - i.e. the `this` of wherever the function was **defined**. "The value of `this` inside an arrow function is always inherited from the enclosing scope."
Bound to the same `this` context of where it was created.

Useful for accessing the `this` of the current env, i.e. in callbacks (`map`).

Practically equivalent:
```
function () {
  return this;
}.bind(this);

() => {
  return this;
};
```

The `this` binding can never be overriden!

resources:
[link1](https://stackoverflow.com/questions/34361379/are-arrow-functions-and-functions-equivalent-interchangeable)
[link2](https://stackoverflow.com/questions/32535110/what-are-the-differences-if-any-between-es6-arrow-functions-and-functions-boun)
[link3](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)

**Comment by Holayn** [comment]
# Immutability

Related to functional programming.

Not mutating objects, instead create new copies of them.
Less complicated programs (since don't have to worry about objects mutating)


Naive implementations of immutable data structures and its operations can result in extremely poor performance because new objects are created each time. It is recommended to use libraries for efficient immutable data structures and operations that leverage on structural sharing.


Reads:
- [link](https://github.com/yangshun/front-end-interview-handbook/blob/master/contents/en/javascript-questions.md#explain-the-difference-between-mutable-and-immutable-objects)
- [link](https://medium.com/javascript-scene/master-the-javascript-interview-what-is-functional-programming-7f218c68b3a0)

**Comment by Holayn** [comment]
TIL: https://stackoverflow.com/questions/7744611/pass-variables-by-reference-in-javascript
another reason why modifying params is a bad idea
